punchcut/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+
+/// Type for a glyph identifier.
+pub type GlyphId = read_fonts::types::GlyphId;
+
+/// Setting for specifying a variation by tag and value.
+#[derive(Copy, Clone, Debug)]
+pub struct Variation {
+    /// Tag for the variation.
+    pub tag: Tag,
+    /// Value for the variation.
+    pub value: f32,
+}
+
+impl From<(Tag, f32)> for Variation {
+    fn from(s: (Tag, f32)) -> Self {
+        Self {
+            tag: s.0,
+            value: s.1,
+        }
+    }
+}
+
+impl From<(&str, f32)> for Variation {
+    fn from(s: (&str, f32)) -> Self {
+        Self {
+            tag: Tag::from_str(s.0).unwrap_or_default(),
+            value: s.1,
+        }
+    }
+}
+
+impl From<([u8; 4], f32)> for Variation {
+    fn from(s: ([u8; 4], f32)) -> Self {
+        Self {
+            tag: Tag::new_checked(&s.0[..]).unwrap_or_default(),
+            value: s.1,
+        }
+    }
+}
+
+/// Context for loading glyphs from any available source.
+#[derive(Default)]
+pub struct Context {
+    /// Inner context for loading TrueType outlines.
+    glyf: glyf::Context,
+    /// Internal storage for TrueType outlines.
+    glyf_outline: glyf::Outline,
+    /// Storage for normalized variation coordinates.
+    coords: Vec<NormalizedCoord>,
+    /// Storage for variation settings.
+    variations: Vec<Variation>,
+}
+
+impl Context {
+    /// Creates a new glyph loading context.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a builder for configuring a scaler.
+    pub fn new_scaler(&mut self) -> ScalerBuilder {
+        ScalerBuilder::new(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        font::*,
+        peniko::kurbo::{
+            PathEl::{self, *},
+            Point,
+        },
+        Context, GlyphId, Scaler,
+    };
+
+    fn test_glyph(font: &FontRef, gid: GlyphId, ppem: f32, expected_elements: &[PathEl]) {
+        let mut cx = Context::new();
+        let mut scaler = cx.new_scaler().size(ppem).build(font);
+        let mut elements = vec![];
+        scaler.outline(gid, &mut elements).unwrap();
+        assert_eq!(&elements[..], expected_elements);
+    }
+
+    #[test]
+    fn unscaled() {
+        let font = FontRef::new(read_fonts::test_data::test_fonts::VAZIRMATN_VAR).unwrap();
+        test_glyph(
+            &font,
+            GlyphId::new(3),
+            0.0,
+            // Path elements in unscaled font units
+            &[
+                MoveTo((281.0, 1536.0).into()),
+                LineTo((474.0, 1242.0).into()),
+                LineTo((315.0, 1242.0).into()),
+                LineTo((57.0, 1536.0).into()),
+                ClosePath,
+            ],
+        );
+    }
+
+    #[test]
+    fn scaled_16_ppem() {
+        let font = FontRef::new(read_fonts::test_data::test_fonts::VAZIRMATN_VAR).unwrap();
+        test_glyph(
+            &font,
+            GlyphId::new(3),
+            16.0,
+            // Path elements scaled to 16ppem as computed by FreeType
```

Comment:
From the information here I don't know how to add additional tests, how exactly do I produce the FreeType #'s for an arbitrary glyph? - could go into an md if you like, in which case "see readme" here might make sense

Reply:
I'd very much like to have a separate repo that is built for large scale testing of all fontations code with crates designed to do that and where we can properly document methodology. I believe @cmyr would like the same for the compiler side. 

In the meantime, I'll add some instructions to reproduce.

Reply:
I added a python script for this to `resources/test_fonts/extract_glyphs.py` and updated `rebuild.sh` to call it. This will generate `extracted/<font-name>-glyphs.text` for each font. I documented the format in the [readme](https://github.com/dfrg/fontations/blob/load-glyphs/resources/test_fonts/README.md).

This should make it fairly simple to add new fonts for testing.